### PR TITLE
feat(rust): improve ockam tcp-outlet commands ux output, help, logs, …

### PR DESCRIPTION
**Pull request by nazmulidris**
Fixes: https://github.com/build-trust/codebase/issues/713
Demo video: https://drive.google.com/file/d/1YvSVzeEUhP1lcnu6hqRFqyGgbFKjCNdj/view

### CODE REVIEW

File: implementations/rust/ockam/ockam_api/src/logs/setup.rs [hunk 1]
```diff
@@ -90,7 +90,7 @@ impl LoggingTracing {
             )
         };
         info!(
-            "tracing initialized. {} {}",
+            "tracing initialized. {:#?} {:#?}",
             logging_configuration, tracing_configuration
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/logs/setup.rs]: Perhaps we could split this in two to avoid polluting too much the `info` channel:

```
info!("tracing initialized");
debug!(%logging_configuration, %tracing_configuration);
```

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/create.rs [hunk 2]
```diff
@@ -80,28 +88,38 @@ impl CreateCommand {
 
         let is_finished: Mutex<bool> = Mutex::new(false);
 
+        // If no alias is supplied, then create a random one (with a friendly name).
+        let alias = match &self.alias {
+            None => random_name().to_string(),
+            Some(alias) if alias.is_empty() => random_name().to_string(),
+            Some(alias) => alias.clone(),
+        };
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/create.rs]: We can avoid this block here removing the `Option` and using clap attributes:

```
#[arg(long, display_order = 901, id = "ALIAS", value_parser = alias_parser, default_value_t = random_name(), hide_default_value = true)]
pub alias: String,
```

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs [hunk 3]
```diff
@@ -9,31 +9,38 @@ use ockam_api::nodes::BackgroundNodeClient;
...
 use crate::util::async_cmd;
-use crate::{color, docs, fmt_ok, CommandGlobalOpts, OckamColor, Terminal, TerminalStream};
+use crate::{docs, fmt_ok, CommandGlobalOpts, Terminal, TerminalStream};
+use crate::{node::NodeOpts, terminal::color_primary};
 
 const AFTER_LONG_HELP: &str = include_str!("./static/delete/after_long_help.txt");
+const LONG_ABOUT: &str = include_str!("./static/delete/long_about.txt");
 
-/// Delete a TCP Outlet
+/// Delete a TCP Outlet.
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs]: Just a heads up: clap trims the stops from docstrings (commands' short docs + commands' args help text)

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs [hunk 4]
```diff
@@ -127,9 +134,9 @@ impl DeleteCommandTui for DeleteTui {
         self.terminal()
             .stdout()
             .plain(fmt_ok!(
-                "Outlet with alias {} on Node {} has been deleted",
-                color!(item_name, OckamColor::PrimaryResource),
-                color!(node_name, OckamColor::PrimaryResource)
+                "Outlet with alias {} on node {} has been deleted",
+                color_primary(item_name),
+                color_primary(&node_name)
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs:139]: Something to keep in mind for a separate PR: we should remove the `color!` macro in favor of the `color_x()` functions.

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs [hunk 5]
```diff
@@ -59,11 +61,21 @@ impl ListCommand {
...
-        let list = opts.terminal.build_list(
-            &outlets.list,
-            &format!("Outlets on Node {}", node.node_name()),
-            &format!("No TCP Outlets found on node {}.", node.node_name()),
-        )?;
+        let list: String = {
+            let empty_message = fmt_info!(
+                "No TCP Outlets found on node {}",
+                color_primary(node.node_name())
+            );
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs:68]: To enforce consistent formatting, we could pass a plain string to `build_list` for the `empty_message` argument and always build an `fmt_info` message with it.

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs [hunk 6]
```diff
@@ -59,11 +61,21 @@ impl ListCommand {
...
+            match outlets.list.is_empty() {
+                true => empty_message,
+                false => opts.terminal.build_list(
+                    &outlets.list,
+                    &format!("TCP Outlets on node {}", color_primary(node.node_name())),
+                    &empty_message,
+                )?,
+            }
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs:76]: Similarly, to avoid this match, we should change the `build_list` behavior to not print the header when the list is empty.

With these two changes to the `build_list` function, you wouldn't need to change the client code in any command and have a consistent output.

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/show.rs [hunk 7]
```diff
@@ -94,13 +100,44 @@ impl ShowTui {
...
+        if outlets.is_empty() {
+            tui.terminal()
+                .stdout()
+                .plain(fmt_info!(
+                    "No TCP Outlets found on node {}",
+                    color_primary(node_name.clone())
+                ))
+                .write_line()?;
+            return Ok(());
+        }
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/show.rs]: This is pretty much what the `tui.show()` function does in its first lines. We use that `Tui` trait exactly to avoid repeating this in every show command. Would it be OK to change the message [here](https://github.com/build-trust/ockam/blob/04cf73ab070197a6502e547da66e26adbe063494/implementations/rust/ockam/ockam_command/src/terminal/tui.rs#L26) instead?

File: implementations/rust/ockam/ockam_command/src/terminal/tui.rs [hunk 8]
```diff
@@ -14,7 +16,21 @@ pub trait ShowCommandTui {
...
+                    .plain(fmt_warn!(
+                        "Failed to show {} {}",
+                        Self::ITEM_NAME.singular(),
+                        color_primary(item_name)
+                    ))
+                    .write_line()?;
+            }
+        }
+        Ok(())
+    }
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/terminal/tui.rs:33]: For a separate PR:

I'd propose a different approach here. Since we want to show multiple items with the same structure as when showing a single item, we should change [this line](https://github.com/build-trust/ockam/blob/04cf73ab070197a6502e547da66e26adbe063494/implementations/rust/ockam/ockam_command/src/terminal/tui.rs#L77) to iterate the list and call `show_single`.

Then, remove the `show_multiple` method and add a new one `show_list()` that internally calls the `list_items_names` + `build_list`. That way we can reuse the implemented trait for both `show` and `list` commands.

**Review by adrianbenavides — COMMENTED** [review]
Looks really nice, I love how all the new details in the output! I left some comments on things to use the traits/utilities we have to avoid adding logic to the commands.

File: implementations/rust/ockam/ockam_command/src/terminal/tui.rs (hunk 8 above)
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/terminal/tui.rs]: For a separate PR:

I'd propose a different approach here. Since we want to show multiple items with the same structure as when showing a single item, we should change [this line](https://github.com/build-trust/ockam/blob/04cf73ab070197a6502e547da66e26adbe063494/implementations/rust/ockam/ockam_command/src/terminal/tui.rs#L77) to iterate the list and call `show_single`.

Then, remove the `show_multiple` method and add a new one `show_list()` that internally calls the `list_items_names` + `build_list`. That way we can reuse the implemented trait for both `show` and `list` commands.

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/static/create/long_about.txt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Create a TCP Outlet that runs adjacent to a TCP server. The Outlet unwraps Ockam messages and delivers the raw TCP messages to the TCP server.
+
+You must specify the TCP address of the server, that your Outlet should send raw TCP traffic to. You can also name your Outlet by giving it an alias.
+
+When you create a TCP Outlet, on an Ockam node, running on your local machine, it makes the TCP service available from a worker address, to the corresponding TCP Inlet (see `ockam tcp-inlet`).
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_command/src/tcp/outlet/static/create/long_about.txt]: In other parts, you refer to `TCP server` instead

File: implementations/rust/ockam/ockam_command/src/terminal/tui.rs (hunk 8 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_command/src/terminal/tui.rs]: Lists are supposed to be a less verbose version compared to the single element. If we always assume the same structure we lose this aspect.
Maybe we can have a `show_single(verbose: bool)`?

**Review by davide-baldo — DISMISSED** [review]

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/static/create/long_about.txt (hunk 9 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/static/create/long_about.txt]: great catch, thank you. done!

File: implementations/rust/ockam/ockam_api/src/logs/setup.rs (hunk 1 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_api/src/logs/setup.rs]: done

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/create.rs (hunk 2 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/create.rs]: great catch; done

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs (hunk 3 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs]: great point; i tend to add trailing period, just for readability; i will strip them +1

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs (hunk 4 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/delete.rs:139]: great point; i will make an issue for this

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs (hunk 5 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs:68]: great idea; done

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs (hunk 6 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/list.rs:76]: great catch! done

File: implementations/rust/ockam/ockam_command/src/tcp/outlet/show.rs (hunk 7 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/tcp/outlet/show.rs]: great idea! done

File: implementations/rust/ockam/ockam_command/src/terminal/tui.rs (hunk 8 above)
Comment (nazmulidris) [reply implementations/rust/ockam/ockam_command/src/terminal/tui.rs]: @adrianbenavides - done

@davide-baldo - the mental model here is the the show command can be given 1 item to show. when it is given many items to show, it just shows each of those items, one after the other. the ux for the following use cases should be the same:
- `ock tcp-outlet show foo` && `ock tcp-oulet show bar`
- `ock tcp-ocket show` (then select foo and bar, press enter, and the output should be the same as above)

show / list: show is meant to display lots of info (eg: project, which has lots of info), and list is meant to show summary info (eg: just a few fields from project are displayed).

**Review by adrianbenavides — APPROVED** [review]
🚀

**Comment by nazmulidris** [comment]
> 🚀

Thank you so much for your guidance & help!
